job-server/src/main/scala/spark/jobserver/JobManagerActor.scala

```diff
@@ -158,24 +163,94 @@ class JobManagerActor(daoActor: ActorRef, supervisorActorAddress: String, contex
+          context.stop() // blocking call since actor is already stopping
+      }
+    }
   }
 
   // Handle external kill events (e.g. killed via YARN)
-  private def sparkListener = {
-    new SparkListener() {
-      override def onApplicationEnd(event: SparkListenerApplicationEnd) {
-        logger.info("Got Spark Application end event, stopping job manager.")
-        self ! PoisonPill
-      }
+  private def sparkListener = new SparkListener() {
+    override def onApplicationEnd(event: SparkListenerApplicationEnd) {
+      logger.info("Got Spark Application end event, stopping job manger.")
+      // Note: This message can be dropped if postStop() has already fired due to PoisonPill.
+      // After PoisonPill, new messages cannot be scheduled.
+      self ! SparkContextStopped
     }
   }
 
   private def isKillingContextOnUnresponsiveSupervisorEnabled(): Boolean = {
     !supervisorActorAddress.isEmpty()
   }
 
+  def adhocStopReceive: Receive = {
+    case StopContextAndShutdown =>
+      // Do a blocking stop, since job is already finished or errored out. Stop should be quick.
+      setContextState(ContextStatus.Stopping) {
+        () => jobContext.stop()
+      }
+
+    case SparkContextStopped =>
+      logger.info("Adhoc context stopped. Killing myself")
+      self ! PoisonPill
+  }
+
+  def stoppingStateReceive: Receive = {
+    // Initialize message cannot come in this state
+    // - If already in stopping state then JVM is also initialized. Initialize won't come
+    // - If restarts then whole JVM will restart and we will have a clean state
+    case ContextStopScheduledMsgTimeout(stopRequestSender) =>
+      logger.warn("Failed to stop context within in timeout. Stop is still in progress")
+      stopRequestSender ! ContextStopInProgress
+
+    case SparkContextStopped =>
+      logger.info(s"Context $contextId stopped successfully")
+      val (stopContextSender, stopContextTimeoutMsgHandler) = stopContextSenderAndHandler
+      stopContextTimeoutMsgHandler.foreach{ handler =>
+        handler.isCancelled match {
+          case true =>
+            // The response to stop request already sent. No need to send any response back.
+          case false =>
+            stopContextTimeoutMsgHandler.foreach(_.cancel())
+            stopContextSender.foreach(_ ! SparkContextStopped)
+        }
+      }
+      self ! PoisonPill
+
+    case StopContextAndShutdown =>
+      logger.info("Context stop already in progress")
+      sender ! ContextStopInProgress
+
+    case StartJob(_, _, _, _, _) =>
+      logger.warn("Tried to start job in stopping state. Not doing anything.")
+      sender ! ContextStopInProgress
+
+    case RestartExistingJobs =>
+      // This message is sent after a watch is added. Terminated will be raised on SJS Master
+      logger.warn("No point in restarting existing jobs as context is in stopping state." +
+        " Will be killed automatically")
+
+    case GetContexData => handleGetContextData
```

Comment:
I think this is a fine pattern to avoid duplicating code but there is another possibility, which is to use a common `PartialFunction` block in multiple receives.  For example:

    val commonHandlers: PartialFunction[....] = {
      case GetContextData => handleGetContextData  // or put original code here
      case KillJob..... handleKilJob
    }

    def stoppingStateReceive: Receive = {....} orElse commonHandlers